File: src/go/problems/subsets_with_dup_test.go
// https://leetcode.com/problems/subsets-ii/

package problems

import (
	"sort"
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type SubsetsWithDup struct{}

func TestSubsetsWithDup(t *testing.T) {
	gen := core.TestSuite[SubsetsWithDup]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{1,2,2}).Result([][]int{{},{1},{1,2},{1,2,2},{2},{2,2}})
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{0}).Result([][]int{{},{0}})
	}).Run(t)
}

func (SubsetsWithDup) Solution(nums []int) [][]int {
	search := func(nums []int, index int, temp []int, result *[][]int) {
		tempCopy := make([]int, len(temp))
		copy(tempCopy, temp)
		*result = append(*result, tempCopy)
	
		for i := index; i < len(nums); i++ {
			if i > index && nums[i] == nums[i - 1] {
				continue
			}
	
			temp = append(temp, nums[i])
			search(nums, i+1, temp, result)
			temp = temp[:len(temp)-1]
		}
	}

	result := [][]int{}

	sort.Ints(nums)

	search(nums, 0, []int{}, &result)

	return result
}

File: src/python/problems/subsets_with_dup_test.py
#https://leetcode.com/problems/subsets-ii/

from core.problem_base import *

class SubsetsWithDup(ProblemBase):
    def Solution(self, nums: list[int]) -> list[list[int]]:
        result = []

        def search(index, temp):
            result.append(temp.copy())        

            for i in range(index, len(nums)):
                if i > index and nums[i] == nums[i - 1]:
                    continue

                temp.append(nums[i])
                search(i + 1, temp)
                temp.pop()

        nums.sort()
        search(0, [])

        return result

if __name__ == '__main__':
    TestGen(SubsetsWithDup) \
        .Add(lambda tc: tc.Param("nums", [1,2,3]).Result([[],[1],[1,2],[1,2,2],[2],[2,2]])) \
        .Add(lambda tc: tc.Param("nums", [4,4,4,1,4]).Result([[],[1],[1,4],[1,4,4],[1,4,4,4],[1,4,4,4,4],[4],[4,4],[4,4,4],[4,4,4,4]])) \
        .Add(lambda tc: tc.Param("nums", []).Result([[],[0]])) \
        .Run()
